### DISCUSSION TITLE
Visual GUI builder

### DISCUSSION THREAD

**Discussion by whiteEelsYikes** [beeware/toga#2315] [Ideas]
可能翻译会出现一些问题，但我并不会英语，如果官方能留意到这条讨论，是一件十分荣幸的事。

我不久前为了构建一个控制物联网设备的py的app发愁，偶然发现了这个库，但是文档并没有指出该库那些位置有那些控件，这些控件的具体api的相关参数等等也没有进行说明，我也不是很想直接翻阅源代码，于是幻想着toge能否有一个类似pyqt那边的qt设计师那样的可视化编程程序。

我大概的浅读了toge的文档，我发现其和qt有是地方是类似的思想（或者是api和一些项目结构设计类似），不知道能不能将qt设计师生成的 **.ui** 文件经过编译转换为符合toge要求的py程序，这样做也方便我在使用中学习toge的结构和使用方式，也非常利于我基于beeware构建自己的应用程序。

如果beeware或者说toge有类似pyqt的qt设计师那种可视化编程程序，我相信beeware一定会更突出，而我也不会在考虑其他框架作为我使用py实现gui程序的首选。

---
Translation via Google Translate:

There may be some translation problems, but I don't speak English. It would be a great honor if the official could pay attention to this discussion.

Not long ago, I was worried about building a py app to control Internet of Things devices, and accidentally discovered this library. However, the document did not indicate which controls are located in the library, and the specific API related parameters of these controls were not explained. I didn't really want to read the source code directly, so I fantasized about whether toge could have a visual programming program similar to the qt designer on pyqt.

I briefly read the toge document and found that it has similar ideas to qt (or the API is similar to some project structure designs). I wonder if the .ui file generated by qt designer can be compiled. Convert it to a py program that meets the requirements of toge. This also facilitates me to learn the structure and usage of toge during use, and is also very helpful for me to build my own application based on beeware.

If beeware or toge has a visual programming program similar to pyqt's qt designer, I believe beeware will be more prominent, and I will not consider other frameworks as my first choice for using py to implement gui programs.

**Comment by freakboy3742** [comment]
Toga does not have a visual designer similar to Qt Designer. We hope to eventually develop a tool like this. At present, our focus is on developing the widgets that a designer app would use.

---

翻譯自Google翻譯

Toga 沒有類似 Qt Designer 的視覺設計器。 我們希望最終能開發出這樣的工具。 目前，我們的重點是開發設計器應用程式將使用的小部件。

**Comment by hyuri** [comment]
Just an idea:
I haven't tried yet, but it might be worth trying to design a GUI in one of the existing GUI designers, like Qt Designer, and asking ChatGPT or similar to convert the .ui code to Toga.

**Reply by freakboy3742** [reply]
You have a *lot* more confidence in ChatGPT than I do. :-)

**Reply by hyuri** [reply]
One can hope :D

**Comment by rexzhang** [comment]
When I used to do cross-platform development with wxWidgets, my favorite tool was https://github.com/wxFormBuilder/wxFormBuilder . It would be good to have a similar tool in the togo ecosystem.

**Comment by tritium21** [comment]
The 1st step in supporting a form builder/designer of some kind is to develop a system in toga to use data to auto-generate a layout.  Traditionally that is done with XML, but I think it could be done much more easily with a dict (and therefor be serialized with any format you want.)  You need to do this even if you intend to make a designer directly generate code - the designer would need this representation internally.  To do this, the API needs to be stable.

The 0th step, therefor, is to get Toga to a (near) finished state before you do any of that.

**Comment by eadmaster** [comment]
> The 1st step in supporting a form builder/designer of some kind is to develop a system in toga to use data to auto-generate a layout.

the old [NetBeans GUI Builder (Codenamed Matisse) swings, JFC ...](https://www.youtube.com/watch?v=Y5RQ6Tk5hBM) was able to generate Java stub classes directly, without the need of a markup language.

I've always found it more intuitive to use than any [custom-made language like the one kivy uses](https://kivy.org/doc/stable/guide/lang.html).

**Reply by freakboy3742** [reply]
Sure - but the complication is reversing that relationship. A GUI code *generator* is of limited utility unless you can modify a generated app after it has been initially generated. If you generate code, the end-developer is going to modify that code - and then the GUI tool needs to parse and interpret those changes. That's the primary benefit of a markup language - it decouples the design from the actual code.

I haven't used the Netbeans GUI builder, but if it's more than a GUI generator, then I'm betting they've imposed some fairly hard restrictions on what you can and can't touch once the code has been generated.

**Reply by eadmaster** [reply]
i've checked one of my old project, and indeed [Matisse was using an xml-based markup internally](https://stackoverflow.com/questions/263959/stand-alone-netbeans-compatible-form-editor), but i never had to edit it directly.

Also you were free to change the auto-generated class/method names if you did not want to track them in the designer tool, but i've found no need to do that. On the flipside i've found consistent naming helped code readability. 

So in the end i guess my bottomline is: any markup is fine as long as i don't have to use it directly and write the bindings manually. :-)

**Reply by freakboy3742** [reply]
To be clear - Toga has no intention of switching to a DSL as the primary mode of GUI creation (in the style of Kivy or Qt). This is a Python toolkit, and we will always endeavour to make the process of writing an app in nothing but Python code as straightforward as possible. A markup language would always be a secondary approach, complimenting rather than replacing the pure-code approach.
